### ISSUE TITLE
Disable submodules via configuration option set to false - [merged]

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1882]
_Merges disable-submodules -> master_

Sibling MR: https://earth.bsc.es/gitlab/ces/autosubmit4-config-parser/-/merge_requests/3

Closes #1130

**Comment by kinow** [comment]
requested review from @dbeltrankyl

**Comment by kinow** [comment]
Sorted imports :point_up:

**Comment by kinow** [comment]
Only change here, but moved the whole code block below, thus the big diff here, but feel free to ignore the changes below.

**Comment by kinow** [comment]
Added types so others know what's the possible types of the `git_project_submodules`.

**Comment by kinow** [comment]
added 1 commit

<ul><li>e43fe805 - Docs for disabling submodules</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/394/diffs?diff_id=19265&start_sha=d6a1a59637257dce7f1acd1ea230406749c591f4)

**Comment by kinow** [comment]
Rendered docs:

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/870053ca0e2ee66d6a0225b9d458179d/image.png)

(Later I think we can re-organise the docs and find a place that is easier for users to locate this info -- for now they can search for submodules.)

**Comment by kinow** [comment]
added 1 commit

<ul><li>8a01870e - Add unit tests, remove unreachable code</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/394/diffs?diff_id=19284&start_sha=e43fe805a55c5972719817577348f46eb250a2e5)

**Comment by kinow** [comment]
Added tests, removed a part of the code that is never reached. Should be ready for review now, @dbeltrankyl .

**Comment by kinow** [comment]
This last one is the test that verifies the new behavior. The others were added to make sure the previous behavior was kept (the best I could do, there are some if/else's in the code I could not reach within one day of development, but we can improve later -- these are the first unit tests for this code).

**Comment by kinow** [comment]
Added this here, and had to move the code one indentation level more.

Also, see line below where I replaced `list.__len__()` by `len(list)`.

**Comment by kinow** [comment]
marked this merge request as **ready**

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 4, 2023, 09:56

Commented on [autosubmit/git/autosubmit_git.py line 248](https://github.com/BSC-ES/autosubmit/compare/df376914acfa72c3399b6efe92b96b75344c408d..b25e752bf469811b0e47ef972eef02539ab8d258#diff-d7b47b8637947d763001053645b70febR248)



Hello Bruno, this line has a format without {0} is it intended?

**Comment by kinow** [comment]
Hmmm, I think it's really pending the parameter reference. I can remove it in a bit (reviewing the optimizations MR). Let me know if you found anything else wrong with this change or if that's all, and I will update it in some minutes/hours :thumbsup: Thanks Dani!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 4, 2023, 10:02

Commented on [test/unit/gui/test_autosubmit_git.py line 134](https://github.com/BSC-ES/autosubmit/compare/df376914acfa72c3399b6efe92b96b75344c408d..b25e752bf469811b0e47ef972eef02539ab8d258#diff-62ac574c5474be46b1f33ee11938ca0fR134)



Hello Bruno,

Thanks for adding the test!

Just a question: 

I see that the autosubmit_git function is called twice in this test. Oddly, the coverage report is unaware of that ( I see red marks in all the function indicating that it is not covered which seems to be false) 

Do you know why?

**Comment by kinow** [comment]
Ah, maybe we need to enable the coverage in the repository like we did in Autosubmit. I don't recall doing that for this one. Good catch. Will update the commit with the coverage settings :sweat_smile: Well spotted.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 4, 2023, 10:05

Commented on [autosubmit/git/autosubmit_git.py line 248](https://github.com/BSC-ES/autosubmit/compare/df376914acfa72c3399b6efe92b96b75344c408d..b25e752bf469811b0e47ef972eef02539ab8d258#diff-d7b47b8637947d763001053645b70febR248)



Hello, 

I think the code itself is correct. Thanks, Bruno! just the question about the coverage

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 4, 2023, 10:06

Commented on [autosubmit/git/autosubmit_git.py line 248](https://github.com/BSC-ES/autosubmit/compare/df376914acfa72c3399b6efe92b96b75344c408d..b25e752bf469811b0e47ef972eef02539ab8d258#diff-d7b47b8637947d763001053645b70febR248)



Also, I see that it is being merged in master, 

I created the 4-0.XXX branch two weeks ago, I think that there were more merges into master, so I will just rebase that branch again

**Comment by kinow** [comment]
Huh, I thought this was the merge request in `autosubmit4-config-parser`, sorry. This one was supposed to have coverage… hmmm :thinking:

**Comment by kinow** [comment]
> I think the code itself is correct. Thanks, Bruno! just the question about the coverage

Alright!

>I created the 4-0.XXX branch two weeks ago, I think that there were more merges into master, so I will just rebase that branch again

:thumbsup:

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jan 31, 2024, 14:33

Commented on [test/unit/gui/test_autosubmit_git.py line 134](https://github.com/BSC-ES/autosubmit/compare/df376914acfa72c3399b6efe92b96b75344c408d..b25e752bf469811b0e47ef972eef02539ab8d258#diff-62ac574c5474be46b1f33ee11938ca0fR134)



Hello @kinow ,

I think this one was done, but there was a weird thing about the coverage, no? 

Did you find the source of the issue? Anyway, it doesn't seem important to the feature functionality so we can merge it if you want

**Comment by kinow** [comment]
Huh, just did this

```bash
python3 -m 'nose' --exclude=regression  --verbosity=3 test/unit --with-coverage --cover-package=autosubmit --cover-inclusive --cover-xml --cover-xml-file=test/coverage.xml 2>&1 | grep submodule
```

And that last `grep` didn't find my test ? :dizzy_face: Let me check this to see if I forgot something...

**Comment by kinow** [comment]
added 5 commits

<ul><li>8a01870e...1e37bc03 - 4 commits from branch <code>master</code></li><li>f724882d - Disable submodules via configuration option set to false</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/394/diffs?diff_id=20720&start_sha=8a01870e9221c4ed8cff9ab64770036786f47466)

**Comment by kinow** [comment]
I was missing a `__init__.py` in the new `gui` test directory :face_palm: Also found a couple of mypy type hint missing imports. Fixed that, tested everything, rebased, squashed :thumbsup: Now wait for CI...

**Comment by kinow** [comment]
Tests failing on Python 3.7 on the CICD worker VM, but passing on my laptop with 3.9.16. Investigating...

**Comment by kinow** [comment]
added 1 commit

<ul><li>09bb0ab6 - Disable submodules via configuration option set to false</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/394/diffs?diff_id=20761&start_sha=f724882de1c497e5023da74646d7b70226f42c52)

**Comment by kinow** [comment]
Alright, I had a property missing in `as_conf`, so I was getting a `str.join(str, Mock, str)`, which apparently works on Py3.9+mock, but not on 3.7. Fixed, tested locally. Pushing the change now...

**Comment by kinow** [comment]
added 1 commit

<ul><li>6230cce0 - Disable submodules via configuration option set to false</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/394/diffs?diff_id=20765&start_sha=09bb0ab685d723b2af3caa57327087dabd9a0b12)

**Comment by kinow** [comment]
There! Coverage fixed, @dbeltrankyl . Thanks for pointing about the missing coverage, Dani! Fixed a bug in the test in Py3.7 and also found out we were not running this test due to the missing `__init__.py`. Care to have another look whenever you have time (low priority), please?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 5, 2024, 09:14

approved this merge request

**Comment by kinow** [comment]
added 1 commit

<ul><li>21ec9f34 - Disable submodules via configuration option set to false</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/394/diffs?diff_id=20843&start_sha=6230cce0d2184260e6ba985836888b64f4adcb41)

**Comment by kinow** [comment]
Hi @dbeltrankyl , last week I got the tests running on GitLab again, and also fixed one test that failed.

Today I was looking at this change before merging (thanks for re-reviewing/approving, BTW :thumbsup:), when I remembered why that test was failing; this change requires a modification in the autosubmit4configparser too, so that it handles when users specify `False` in the list of submodules: https://earth.bsc.es/gitlab/ces/autosubmit4-config-parser/-/merge_requests/3

I updated the commit to have the test failing again, but once that autosubmit4configparser MR is reviewed/merged I can trigger a new run here and then tests should pass and we can merge it (I just tested locally).

The autosubmit4configparser MR also adds the initial unit tests to the project, which we should be able to run automatically after today's discussion with @fbeninca about GitLab workers.

Can you take a look at that MR too, whenever you have time, please? I will switch to the workflow optimizations now.

Thanks!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 4, 2024, 13:09

resolved all threads

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 4, 2024, 13:21

Hello @kinow ,

Is this ready to be merged? From my side, I would say yes

Or was it related to this one? https://earth.bsc.es/gitlab/ces/autosubmit4-config-parser/-/merge_requests/2

**Comment by kinow** [comment]
It was blocked by this one https://earth.bsc.es/gitlab/ces/autosubmit4-config-parser/-/merge_requests/3, which is now merged.

Let me rebase it. If the build passes it should be ready to be merged. Thanks for the bump Dani!

**Comment by kinow** [comment]
added 5 commits

<ul><li>21ec9f34...f32a168b - 4 commits from branch <code>master</code></li><li>3ba54a69 - Disable submodules via configuration option set to false</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/394/diffs?diff_id=22198&start_sha=21ec9f342a33f90cdc2c0a0020e9dab264cd9ae0)

**Comment by kinow** [comment]
I thought my change hadn't worked as the pipeline failed.

>Fetching changes...
>error: failed to write new configuration file /home/gitlab-runner/builds/o7zBmX1g/0/gitlab/es/autosubmit.tmp/git-template/config.lock
>Cleaning up project directory and file based variables 00:00
>ERROR: Job failed: exit status 1

Seems like some issue with the worker (cc @fbeninca https://earth.bsc.es/gitlab/es/requests/-/issues/2370)

**Comment by kinow** [comment]
Your logs rework merge request has the same issue, @dbeltrankyl , !401 . I'm trying to kick GitLab to fix that, do you know how to fix that one? Not sure if you've seen it before? ...

**Comment by kinow** [comment]
Tried clearing runner caches, but no luck. Looks like all our merge requests are now failing.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/852a9d2508bf439a1e569bf59b137ee3/image.png)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 17, 2024, 17:50

This had issues with the worker; I guess this will be rebased and merged after !445?

**Comment by kinow** [comment]
added 285 commits

<ul><li>3ba54a69...df376914 - 284 commits from branch <code>master</code></li><li>b25e752b - Disable submodules via configuration option set to false</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/394/diffs?diff_id=30022&start_sha=3ba54a69a36fc739adc549c50522132467547ef9)

**Comment by kinow** [comment]
Rebased, now waiting for CICD to pass, @dbeltrankyl . The merge request was blocked by pending changes in the config parser, which have already been merged. So if the tests pass this should be ready to be merged.

**Comment by kinow** [comment]
mentioned in commit c418eac60fd752f4a643a1b9d560cafb6c57d2fd
